File: models/refund.go
package models

import (
	"time"

	"go.mongodb.org/mongo-driver/bson/primitive"
)

type RefundStatus string

const (
	RefundStatusPending RefundStatus = "pending"
	RefundStatusSigned  RefundStatus = "signed"
	RefundStatusSuccess RefundStatus = "success"
	RefundStatusInvalid RefundStatus = "invalid"
)

type Refund struct {
	ID                    *primitive.ObjectID `json:"_id,omitempty" bson:"_id,omitempty"`
	OriginTransaction     *primitive.ObjectID `json:"origin_transaction" bson:"origin_transaction"`
	OriginTransactionHash string              `json:"origin_transaction_hash" bson:"origin_transaction_hash"`
	Recipient             string              `json:"recipient" bson:"recipient"`
	Amount                uint64              `json:"amount" bson:"amount"`
	TransactionBody       string              `json:"transaction_body" bson:"transaction_body"`
	Signatures            []Signature         `json:"signatures" bson:"signatures"`
	Transaction           *primitive.ObjectID `json:"transaction" bson:"transaction"`
	TransactionHash       string              `json:"transaction_hash" bson:"transaction_hash"`
	Status                RefundStatus        `json:"status" bson:"status"`
	CreatedAt             time.Time           `bson:"created_at" json:"created_at"`
	UpdatedAt             time.Time           `bson:"updated_at" json:"updated_at"`
}

File: e2e/tester/src/types/refund.ts
// Import the required dependencies for primitive types and date handling
import { ObjectId } from 'mongodb';
import { Hex } from 'viem';

// Assuming Signature type is imported from a local file
import { Signature } from './message';

// Define the RefundStatus type
export type RefundStatus = 'pending' | 'signed' | 'broadcasted' | 'success' | 'invalid';

// Define the Refund type
export type Refund = {
  _id?: ObjectId;
  origin_transaction: ObjectId;
  origin_transaction_hash: Hex;
  recipient: string;
  amount: number;
  transaction_body: string;
  signatures: Signature[];
  transaction?: ObjectId;
  sequence?: number;
  transaction_hash: Hex;
  status: RefundStatus;
  created_at: Date;
  updated_at: Date;
};

export const CollectionRefunds = 'refunds';
